spec/consensus/reactor/implementation.md

```diff
@@ -0,0 +1,212 @@
+
+> **TODO** .   
+> Discuss each kind of message and make the previous itemize a proper argument, which will allow showing that $c$ exists, even if high.
+
+
+## P2P-Only nodes
+[REQ-GOSSIP-CONS-SUPERSESSION.2] implies that the messages being broadcast by the process itself and those being forwarded must be limited.
+In Tendermint BFT this is achieved by virtue of only validators broadcasting messages and the set of validators being always limited.
+
+Although only validators broadcast messages, even non-validators (including sentry nodes) must deliver them, because: 
+* only the nodes themselves know if they are validators,
+* non-validators may also need to decide to implement the state machine replication, and,
+* the network is not fully connected and non-validators are used to forward information to the validators.
+
+Non-validators that could support applications above them may be able to inform GOSSIP about superseded messages (for example, upon decisions).
+
+Non-validators that are deployed only to facilitate communication between peers, that is, P2P only nodes that implement GOSSIP but not the CONS, still need to be provided with a supersession operator in order to limit buffering.
+
+
+All nodes currently run Tendermint BFT, but desire to have lightweight, gossip only only, nodes has been expressed, e.g. in [ADR052](#references)
+
+
+## [DEF-BROADCAST]
+1. BMsgs is implemented by the gossip routines encountering certain conditions that will trigger the broadcast of messages.
+1. Broadcast messages are put in outgoing queues for current connections.
+1. If the condition persists, then they are put in the queue again on later iterations, for certain messages.
+1. If the condition persists, then they are put in the queues for new connections in later iterations.
+1. If the condition does not persist, then some superseding messages must have been sent and the superseded one may be dropped (same as remaining in BMsgs but never being actually sent again on the network).
+
+1. Superseding is not explored to drop superseded messages from the outgoing queues.
+
+## [LOCAL_CAST.1]
+For each of the neighbors of the node, looping go-routines continuously evaluate the conditions to send messages to other nodes.
+If a message must be sent, it is enqueued for transmission using TCP and will either be delivered to the destination or the connection will be dropped.
+New connections reset the state of Communication layer wrt the newly connected node (in case it is a reconnection) and any messages previously sent (but possibly not delivered) will be resent if the conditions needed apply. If the conditions no longer apply, it means that the message has been superseded and need no be retransmitted.
+
+1. Every message added to BCast is enqueued or will be enqueued later.
+2. Every message enqueued will be delivered using TCP or the connection will break.
```

Comment:
It's possible that an "enqueued" message is not delivered, and the corresponding connection is not broken, for an indefinite period of time. Defining a timeout at the application layer can break the connection at a certain point, but even in that case the message may still be delivered to the remote host. Hopefully I'm just preaching to the choir and this is all well understood 🙃 but if not please speak up as I guess this is important?

Reply:
The point here is that if the message is not delivered once enqueued, it will be enqueued again, eventually.
And eventually it will be delivered or it will not be needed anymore.

Reply:
I understand these are the expectations at the application layer, but how are those expectations guaranteed? What re-enqueues messages that aren't delivered? What determines if an un-delivered message is no longer necessary and removes it from the queue?

Reply:
> What determines if an un-delivered message is no longer necessary
The analysis of the current state of the node. If the node has moved to heigh X, there is no point in sending messages for height X-2, for example.
> and removes it from the queue?
There is no need to remove it from the queue. Just don't enqueue it again.

Reply:
OK, I think I get it. This all assumes that delivery success/failure is knowable from the perspective of the sender. That's a property asserted by Gossip Communication (which is probably where you're speaking from) but not actually provided by e.g. TCP connections over a physical network (which is where I've been speaking from).

Reply:
Actually, the sender does not need to know, as long as it keep insisting and sending any messages that might still be useful. TCP cannot guarantee the delivery, but if you insist (reconnect if needed), in practice, stable periods will happen and deliver will succeed. Again, the sender may never know when this happens, so it needs to keep retransmitting. 

This is one place where theory does not meet practice. We write algorithms and prove that they will progress after GST, but we cannot guarantee that there will be a GST. However, as long GST is required only for liveness, not safety, and as long as the algorithm may still make progress during small stability periods, we get progress even if GST never arrives. This is the case with TM.

Reply:
The problem is that it's not possible for a sender to know if a given message has actually been received by its destination. An acknowledgement (ACK) is most likely a true-positive signal that the message was received, but a NAK is not a true-negative signal that the message was not received. That means the application can't treat a NAK as a signal to delete a message from its transmit queue, and can only stop trying to transmit a message once an alternative, out-of-band signal is received, that indicates that message is no longer useful. I guess that's the committed block height received over consensus. That's fine. But what happens if that latest block height data is delayed indefinitely? It means that GST is unbounded. You can classify such a node as incorrect, and thereby remove it from consideration of your system model, but that classification relies on information that is not available to the node itself, or (potentially) other nodes that also did not receive the latest block height data. So we're back to square one in terms of correctness being a relative measure. Again, this is all kind of fine, in terms of correctness! The only consequence I'm trying to point out is that GST is not guaranteed by the protocol, and can't be assumed by any implementation.

Reply:
Please read my previous message again, where I clearly stated that "...the sender does not need to know, as long as it keep insisting and sending any messages that might still be useful"  and "the sender may never know when this happens, so it needs to keep retransmitting."

> can only stop trying to transmit a message once an alternative, out-of-band signal is received, that indicates that message is no longer useful.
It is not out of band, it is in the consensus protocol, which is the reason for sending the messages.

> It means that GST is unbounded.
This phrase does not make sense.

> You can classify such a node as incorrect, and thereby remove it from consideration of your system model, but that classification relies on information that is not available to the node itself, or (potentially) other nodes that also did not receive the latest block height data.
I am not the one that wants to classify an incommunicado node as faulty.

> So we're back to square one in terms of correctness being a relative measure. 
No we are not. A correct process is a correct process. A faulty process is a faulty process. The same applies to links, independently.

> GST is not guaranteed by the protocol...
Indeed it is not. It is assumed of the environment. FLP implies that if you want to prove that your algorithm eventually makes progress and terminates then you need to extend your asynchronous system somehow, be it with randomization oracles, perfect failure detectors or, in the case of tendermint, GST. 

> and can't be assumed by any implementation.
That's exactly what you have to do. You write you algorithm so that it makes progress when GST arrives and expect that it will progress even if it is not really GST, but just some stable period. You cannot guarantee it will be enough, but it is best you can do, given FLP.

Reply:
> A correct process is a correct process. A faulty process is a faulty process. The same applies to links, independently.

I don't understand how this is true.

Some process A can be correct from the perspective of an observer process X, and incorrect from the perspective of a different observer process Y. (If this is not evident, let me know, I can provide relevant examples.) Is that process A correct or incorrect? AFAICT, it is neither; there is no well-defined answer, only relative perspectives.

Similarly, the link between A and X can be correct from the perspective of A, and incorrect from the perspective of X. Is that A-X link correct or incorrect? AFAICT, same outcome as above.

It's entirely possible I'm missing something here. What am I missing?

--

> FLP implies that if you want to prove that your algorithm eventually makes progress and terminates then you need to extend your asynchronous system somehow, be it with randomization oracles, perfect failure detectors or, in the case of tendermint, GST.

I think the problem I'm observing here is that this invariant -- that the algorithm will eventually make progress -- is fine as an assumption for a system model, but not fine as an assumption for a real-world implementation. Basically, FLP asserts that it is not actually possible to for an algorithm to make progress and terminate, in any real-world context. Following: that GST is not guaranteed, or guarantee-able, in any real-world context. Does that mean that an algorithm does not necessarily make progress and/or terminate? Yes! It does mean that! Randomization oracles don't solve this problem; perfect failure detectors are not feasible in any real-world network.